meson: add support for testsuite

eli-schwartz opened the pull request:
As with CMakeLists.txt, this is disabled by default.

- I have not implemented the examples/ directory because it uses some output regex feature to run the tests, and meson doesn't have any similar feature, so that would probably require some kind of wrapper program that runs the test binary and compares output.
- I have not enabled doxygen documentation building, because it doesn't seem very useful in the same way that running tests under meson is; also, setting the output directory to be in the builddir requires either a python script to edit Doxyfile, or for Doxyfile to set `@VARNAME@` that can be replaced (breaking usage without first running cmake or meson) or for me to stop being lazy and finally implement doxygen in a meson module. :D
- I have not enabled CUDA tests, because I'm not sure if those are important?
- Testing `CLI11_SINGLE_FILE` doesn't seem very useful, because it should behave exactly the same anyway, modulo speed of e.g. deduplicated includes 
- There are some features which I didn't bother to implement, because meson supports it out of the box:
  - sanitizers can be configured with meson using `-Db_sanitize=`, possible values include `address`, `thread`, `undefined`, `memory`, `address,undefined`. AFAIK this functionality is not tested for cmake in CI anyway.
  - a clang-tidy command is automatically set up by meson if the `clang-tidy` program is installed.
  - code coverage can be configured with meson using `-Db_coverage=true`

eli-schwartz committed:
meson: add support for testsuite

eli-schwartz committed:
meson: fix deprecation warning for run_command with unchecked return code

This would implicitly default to false, so if something bizarre happened
and the command errored out, meson would consider that fine. Now meson
emits a warning about this deprecated legacy behavior, suggests that it
will eventually change, and, most importantly, prevents a warning-free
build.

Suppress the warning by manually specifying the sensible behavior, which
is to fail on errors.

henryiii commented on tests/meson.build:
```diff
@@ -0,0 +1,78 @@
+    ['TrueFalseTest', {}],
+    ['OptionGroupTest', {}],
+    # multi-only
+    ['TimerTest', {}]
```
```suggestion
    ['TimerTest', {}],
```

eli-schwartz replied:
Oops, I added that at the last minute and apparently forgot to rerun on modifications.

eli-schwartz replied:
Well, to be more specific I moved link_test_2 into that array instead of having it created standalone on line 18...

Anyway, fixed now.

henryiii replied:
Looks like there's a trailing comma now too, that's good. :)

eli-schwartz replied:
Indeed, and I regularly give thanks that Meson was able to avoid "the JSON mistake" there. :D

eli-schwartz committed:
meson: download catch2 on demand if a system version is unavailable

Produced by running `meson wrap install catch2` and checking the results
into git.

No modifications to the build files are expected; this makes use of
https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section

pre-commit-ci[bot] committed:
style: pre-commit.ci fixes

henryiii committed:
ci: add meson build to the CI

henryiii committed:
ci: meson doesn't depend on ninja or or have a ninja extra

henryiii commented on tests/meson.build:
```diff
@@ -0,0 +1,78 @@
+catch2 = dependency('catch2')
```
Just for my understanding, this will download catch only if not found? And any version can be found, but 2.13.7 will be downloaded, correct?

eli-schwartz replied:
Yes. Based on the setting of `--wrap-mode=default` (you can set it to `=forcefallback` if you'd prefer to download it no matter what) it will:
- try to find any version using pkg-config
- try to find any version using cmake (by running a crafted CMakeLists.txt, seeing if find_package() works, and extracting that info using the trace parser)
- download the pinned version available as a meson wrap

henryiii committed:
ci: minor cleanup to Meson job
